samples: fs: littlefs sample program with block devices support

lmajewski opened the pull request:
Up till now, the littlefs was only supporting the flash medium. However,
it would be also welcome to be able to read data from littlefs stored
on block devices - like SD card.

In this case we do use the block API, to read and write data to it.

Two tests are provided - one is similar to the already in-tree littlefs
sample - the value from file is read, incremented and stored again.

Second test is a bit more verbose - the binary manipulation on file
larger than single block and not dividable by block is performed.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

lmajewski commented:
Patch in this PR has been extracted from https://github.com/zephyrproject-rtos/zephyr/pull/40704 to separate block device support on the `littlefs` from its sample program.

nvlsianpu commented:
@lmajewski Can the littlefs sample in the code be extended instead? I think it miss the second test and the block-device configuration only (+ some refactoring).
Rationale for that are:
* littlefs capability are independent from the medium it uses
* less maintenance on one sample than with two

Commit:
fs: littlefs: sample: Replace calls to printk with LOG_PRINTK

Replace calls to printk with LOG_PRINTK to use the logging Zephyr
infrastructure. After this change it would be possible to set the
logging mode (to e.g. reduce the verbosity of output and final size
of the binary).

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: sample: Add test with binary pattern file modification

This commit extends the littlefs test to use binary pattern file.

This file has the 0x55 and 0xAA patterns (to avoid fast wear leveling)
written alternately with special "marks" bytes on each eight position.
Those values are incremented each time the board boots. One can adjust
the size of this file with the TEST_FILE_SIZE define.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: Introduce CONFIG_APP_LITTLEFS_STORAGE_FLASH define

This KConfig define (by default defined) is necessary to extend the
littlefs test functionality to support other test scenarios - for
example running it on device with block device (i.e. SD card) backend.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: Extend sample program to test block devices

Up till now, the littlefs sample program was only supporting the flash
medium. However, it would be also welcome to be able to test reading and
writing data from littlefs stored on block devices - like SD card.

In this case we do use the block API, to read and write data to it.
To achieve this goal - the CONFIG_APP_LITTLEFS_STORAGE_BLK_SDMMC define
has been introduced.

One can change the disk mount point by defining board specific
CONFIG_SDMMC_VOLUME_NAME.

The test for nucleo board equipped with SD card shall be built with:
west build -p always -b nucleo_h743zi ./zephyr/samples/subsys/fs/littlefs \
 -DCONF_FILE=prj_blk.conf

Moreover, the README.rst has been extended to describe running this
test on block devices.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: sample: Move littlefs_flash_erase() to flash mount

The littlefs_flash_erase() function shall be called from flash specific
littlefs_mount function. This allows removal of the stub function
when block device support is added.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: sample: Move listing directory code to its own function

The code for presenting content of the directory, has been moved to the
separate function to be easily reusable.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: sample: Move value increment test to separate function

The code responsible for incrementing the in-file value has been moved
to the separate function for better readability and reusage.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: sample: Move flash initialization code to its own function

The code for potential erasing the flash area has been moved to the
separate function.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
fs: littlefs: sample: Generate fname[12] strings after mount

The littlefs_mount() function sets the mount point for block devices,
so fname[12] shall be generated after calling it.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

lmajewski commented:
@nvlsianpu @Laczen @de-nordic @jfischer-no I've refactored "a bit" the `littlefs` sample program to also support the block devices use case.

lmajewski commented:
@nvlsianpu @Laczen @de-nordic @jfischer-no Any comments regarding this patch set?

nvlsianpu commented on samples/subsys/fs/littlefs/Kconfig:
```diff
@@ -17,6 +17,8 @@ choice
 config APP_LITTLEFS_STORAGE_FLASH
 	bool "Use flash memory backend"
 
+config APP_LITTLEFS_STORAGE_BLK_SDMMC
```
I would not introduce new option but I would make build decision depends on that whether CONFIG_FS_LITTLEFS_BLK_DEV==y instead. This imply that APP_LITTLEFS_STORAGE_FLASH, APP_LITTLEFS_STORAGE_BLK_SDMMC property would be not needed (in source code additional intermediary defines might be usefully).

This will remove some redundancy when configuration need to select the application property and closely linked the littlefs subsystem property.

nvlsianpu commented on samples/subsys/fs/littlefs/src/main.c:
```diff
@@ -119,6 +121,128 @@ static int littlefs_increase_infile_value(char *fname)
+		return rc;
+	}
+
+	rc = fs_close(&file);
```
I recommend to call fs_cloce() always if fs_open() succed before/

nvlsianpu reviewed (commented):
Thanks for this rework. It Looks promising.

lmajewski commented:
@nvlsianpu There is a cleanup patch for the `littlefs` - https://github.com/zephyrproject-rtos/zephyr/pull/41972. 
@Laczen Introduced there new Kconfig define - `FS_LITTLEFS_FMP_DEV` which can be used in the reworked test setup as well as the already present - `FS_LITTLEFS_BLK_DEV`

I will prepare new version of this PR after the https://github.com/zephyrproject-rtos/zephyr/pull/41972 is merged.

nvlsianpu commented:
OK, lets wait for it then.

Laczen commented:
@lmajewski, sorry I had to close that PR,

I don't have the time to work on waste of time change-requests. I was open to discussions on further splitting up into separate files as you might remember. But not to spend my time on something that was absolutely waste.

I will swiftly approve your PR when it comes in.

Laczen commented:
@lmajewski, sorry I had to close that PR,

I don't have the time to work on waste of time change-requests. I was open to discussions on further splitting up into separate files as you might remember. But not to spend my time on something that was absolutely waste.

I will swiftly approve your PR when it comes in.

lmajewski commented:
> @lmajewski OK

@nvlsianpu @Laczen Please review updated patch set for enhancing the `littlefs` sample code.

Laczen reviewed (dismissed):
LGTM, thanks.

lmajewski commented:
@nvlsianpu - Would you find some time to review this change?

de-nordic commented on samples/subsys/fs/littlefs/src/main.c:
```diff
@@ -12,11 +13,249 @@
+		   boot_count, rc);
+
+ out:
+	rc = fs_close(&file);
```
The `rc` gets overwritten here, so in case when we get to `out` due to error and are successfully closing the file, the overall return from function would be success. This does not seem to be correct.

de-nordic commented on samples/subsys/fs/littlefs/src/main.c:
```diff
@@ -12,11 +13,249 @@
+	}
+
+ out:
+	rc = fs_close(&file);
```
Same as https://github.com/zephyrproject-rtos/zephyr/pull/41812/files#r818639317

de-nordic commented on samples/subsys/fs/littlefs/src/main.c:
```diff
@@ -31,132 +270,104 @@ static struct fs_mount_t lfs_storage_mnt = {
-	if (rc >= 0) {
-		printk("\tfn '%s' size %zu\n", dirent.name, dirent.size);
-	}
+	mp->storage_dev = (void *) disk_pdrv;
```
Space after cast.

de-nordic commented on samples/subsys/fs/littlefs/src/main.c:
```diff
@@ -31,132 +270,104 @@ static struct fs_mount_t lfs_storage_mnt = {
 
-	while (rc >= 0) {
-		struct fs_dirent ent = { 0 };
+	lsdir(mp->mnt_point);
```
You are not checking result of the function here so you want to proceed whether this failed or not?

de-nordic commented on samples/subsys/fs/littlefs/src/main.c:
```diff
@@ -12,11 +13,249 @@
+	LOG_PRINTK("\nListing dir %s ...\n", path);
+	for (;;) {
+		/* Verify fs_readdir() */
+		res = fs_readdir(&dirp, &entry);
+
+		/* entry.name[0] == 0 means end-of-dir */
+		if (res || entry.name[0] == 0) {
+			break;
+		}
```
There is no log on `res != 0` here, so there is no difference between empty dir and error, as you do in other cases. If you rely on `fs_readdir` logging error here, this is fine just make sure VFS logs are on.

lmajewski replied:
I've added some extra `LOG_ERR()` function to present the error.

lmajewski commented:
@de-nordic @nvlsianpu @Laczen Please find next revision of this patch set with adjustments proposed by @de-nordic.

de-nordic reviewed (approved):
Looks OK.

nvlsianpu reviewed (approved):
checked once more.
LGTM
